### PR TITLE
Allow iterators to return current position (index) during {{#each}}

### PR DESCRIPTION
**Pull request by Joelkang**
In the `opcodes/lists.ts` positional params only takes PathReferences. Not entirely sure if I should convert it to take any kind of reference, or if my casting to <any> is acceptable.

### CODE REVIEW

**Comment by chadhietala** [comment]
This looks like it will fix `{{#each}}` as well.

**Comment by Joelkang** [comment]
Yea, not 100% sure where I should add tests for this tbh; could use some direction there

**Comment by chadhietala** [comment]
@Joelkang Are you implementing `each-in` in glimmer or Ember? This might be a question for @wycats. You can probably add the tests here https://github.com/tildeio/glimmer/blob/master/packages/glimmer-runtime/tests/updating-test.ts or a better idea may be to make a new file in the runtime tests for this feature.

**Comment by Joelkang** [comment]
I'm implementing it in ember but this is a prerequisite for that. I
actually already have an ember prototype but let me verify with test before
I throw up a PR. Will write tests for this in the meanwhile; thanks for the
pointer

On Friday, May 6, 2016, Chad Hietala <notifications@github.com> wrote:

> @Joelkang <https://github.com/Joelkang> Are you implementing each-in in
> glimmer or Ember? This might be a question for @wycats
> <https://github.com/wycats>. You can probably add the tests here
> https://github.com/tildeio/glimmer/blob/master/packages/glimmer-runtime/tests/updating-test.ts
> or a better idea may be to make a new file in the runtime tests for this
> feature.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tildeio/glimmer/pull/166#issuecomment-217585051>
>


-- 
Cheers,
Joel

File: packages/glimmer-reference/lib/iterable.ts [hunk 1]
```diff
@@ -138,6 +140,10 @@ export class ReferenceIterator {
     this.artifacts = artifacts;
   }
 
+  getPosition(): ConstReference< number | FIXME<'user str to InternedString'>> {
```
Comment (krisselden) [inline packages/glimmer-reference/lib/iterable.ts]: should be a ValueReference

**Comment by krisselden** [comment]
Looks like this just failed because of the typescript version being unpinned.  I submitted a fix https://github.com/tildeio/glimmer/pull/169

File: packages/glimmer-reference/lib/iterable.ts [hunk 2]
```diff
@@ -8,6 +9,7 @@ export interface IterationItem<T> {
 
 export interface AbstractIterator<T, U extends IterationItem<T>> {
   isEmpty(): boolean;
+  getPosition(): number | FIXME<'user str to InternedString'>;
```
Comment (krisselden) [inline packages/glimmer-reference/lib/iterable.ts]: Should this be Opaque?

File: packages/glimmer-reference/lib/iterable.ts [hunk 3]
```diff
@@ -86,18 +91,18 @@ export class IterationArtifacts {
...
     list.append(node);
     return node;
   }
 
-  insertBefore(item: OpaqueIterationItem, reference: ListItem): ListItem {
+  insertBefore(item: OpaqueIterationItem, reference: ListItem, position: number | FIXME<'user string to InternedString'>): ListItem {
     let { map, list, iterable } = this;
 
-    let node = map[item.key] = new ListItem(iterable, item);
+    let node = map[item.key] = new ListItem(iterable, item, position);
```
Comment (krisselden) [inline packages/glimmer-reference/lib/iterable.ts]: why are we accepting a ListItem and a position, and then creating a new ListItem?

File: packages/glimmer-reference/lib/iterable.ts [hunk 4]
```diff
@@ -138,22 +143,27 @@ export class ReferenceIterator {
...
 
     let iterator = (this.iterator = this.iterator || artifacts.iterate());
 
     let item = iterator.next();
+    let position = iterator.getPosition();
 
     if (!item) return null;
 
-    return artifacts.append(item);
+    return artifacts.append(item, position);
```
Comment (krisselden) [inline packages/glimmer-reference/lib/iterable.ts]: I'm not sure why position is sometimes part of item and sometimes not.  Why isn't it like value where the list item is updated?

File: packages/glimmer-test-helpers/lib/environment.ts [hunk 5]
```diff
@@ -96,25 +97,77 @@ class ArrayIterator implements OpaqueIterator {
...
+
+  constructor(kvPairs: KVPair[]) {
+    this.array = kvPairs;
+  }
+
+  isEmpty(): boolean {
+    return this.array.length === 0;
+  }
+
+  getPosition(): string {
```
Comment (krisselden) [inline packages/glimmer-test-helpers/lib/environment.ts]: this feels a little weird.

**Comment by chancancode** [comment]
Great work!

Some suggestions:

As you have identified, the second arg is used for more than just "position"s. I think we should just call it something more generic. I think it would be a good idea to make that a generic "memo" slot that the iterator can use to yield additional information to the template.

The public contract (that Ember uses) becomes:

```typescript
export interface IterationItem<T, U> {
  key: string;
  value: T;
  memo: U;
 }

export interface AbstractIterator<T, U, V extends IterationItem<T, U>> {
  isEmpty(): boolean;
  next(): V;
}

export interface AbstractIterable<T, U, ItemType extends IterationItem<T, U>, ValueReferenceType extends PathReference<T>, MemoReferenceType extends PathReference<U>> {
  iterate(): AbstractIterator<T, ItemType>;

  valueReferenceFor(item: ItemType): ValueReferenceType;
  updateValueReference(reference: ValueReferenceType, item: ItemType);

  memoReferenceFor(item: ItemType): MemoReferenceType;
  updateMemoReference(reference: MemoReferenceType, item: ItemType);
}
```

This allows Ember's iterator to yield whatever it needs for the second arg (and use whatever reference types it sees fit). As far as Glimmer is concerned, it's just another opaque value we pass along to the template.

Once you have that, the rest should just fall into place with relatively little changes.

File: packages/glimmer-reference/lib/iterable.ts (hunk 4 above)
Comment (Joelkang) [inline packages/glimmer-reference/lib/iterable.ts]: I think the reason why did this was that it didn't make sense to me that the position was part of the iterationItem interface since it was meta-item. Because of this I wound up passing everything across the stack. Given both your and Godfrey's comments, I'll make it internal to IterationItem so there's no more verbose passing around

**Comment by Joelkang** [comment]
@chancancode  @krisselden updated per your comments

File: packages/glimmer-reference/lib/iterable.ts [hunk 6]
```diff
@@ -152,9 +160,9 @@ export class ReferenceIterator {
 }
 
 export interface IteratorSynchronizerDelegate {
-  retain(key: InternedString, item: PathReference<any>);
-  insert(key: InternedString, item: PathReference<any>, before: InternedString);
-  move(key: InternedString, item: PathReference<any>, before: InternedString);
+  retain(key: InternedString, item: PathReference<any>, memo: PathReference<any>);
```
Comment (krisselden) [inline packages/glimmer-reference/lib/iterable.ts]: why are these any instead of Opaque?

File: packages/glimmer-test-helpers/lib/environment.ts [hunk 7]
```diff
@@ -80,14 +80,14 @@ import {
   UpdatableReference
 } from "glimmer-object-reference";
 
-type KeyFor = (item: Opaque, index: number) => string;
+type KeyFor = (item: Opaque, index: number | string) => string;
```
Comment (krisselden) [inline packages/glimmer-test-helpers/lib/environment.ts]: why is this changed? seems like memo leaking into KeyFor

File: packages/glimmer-reference/lib/iterable.ts (hunk 6 above)
Comment (chancancode) [inline packages/glimmer-reference/lib/iterable.ts]: Good question. In general the answer is probably "this code predates the introduction of `Opaque`", so it might be time to fix it!

File: packages/glimmer-test-helpers/lib/environment.ts (hunk 7 above)
Comment (chancancode) [inline packages/glimmer-test-helpers/lib/environment.ts]: Since this is only used by the tests, I'm not too sad about it. We can probably make it `type keyFor<T> = (item: Opaque, memo: T) => string;`. Then `ArrayIterator` can have a `KeyFor<number>`, and `ObjectKeysIterator` can have a `KeyFor<string>`.

File: packages/glimmer-runtime/lib/vm/update.ts [hunk 8]
```diff
@@ -1,7 +1,8 @@
 import { Scope, DynamicScope, Environment } from '../environment';
 import { Bounds, clear, move as moveBounds } from '../bounds';
 import { ElementStack, Tracker } from '../builder';
-import { LOGGER, Destroyable, Stack, LinkedList, InternedString, Dict, dict } from 'glimmer-util';
+import { LOGGER, Destroyable, Stack, FIXME, LinkedList, InternedString, Dict, dict } from 'glimmer-util';
```
Comment (chancancode) [inline packages/glimmer-runtime/lib/vm/update.ts]: Some unused stuff here (`FIXME` is what my 👀 could spot, but tslint would tell you for sure)

File: packages/glimmer-test-helpers/lib/environment.ts [hunk 9]
```diff
@@ -146,18 +182,30 @@ class Iterable implements AbstractIterable<Opaque, IterationItem<Opaque>, Updata
       return array.length > 0 ? new ArrayIterator(array, keyFor) : EMPTY_ITERATOR;
     } else if (iterable === undefined || iterable === null) {
       return EMPTY_ITERATOR;
-    } else {
+    } else if (typeof iterable === 'object') {
+       let keys = Object.keys(iterable);
+       let kvPairs = keys.map(key => <KVPair>{ key, value: iterable[key] });
```
Comment (chancancode) [inline packages/glimmer-test-helpers/lib/environment.ts]: Usually you wouldn't need to do an explicit cast like this. If it couldn't figure it out, something is probably wrong.

File: packages/glimmer-test-helpers/lib/environment.ts [hunk 10]
```diff
@@ -96,17 +96,53 @@ class ArrayIterator implements OpaqueIterator {
...
 
-    return { key, value };
+    return { key, value, memo };
+  }
+}
+
+type KVPair = {
+  key: string;
+  value: Opaque;
+};
```
Comment (chancancode) [inline packages/glimmer-test-helpers/lib/environment.ts]: You could write

```typescript
interface KVPair {
  key: string;
  value: Opaque;
}
```

(Or the iterator could just take an array of keys and an array of values, since it's all internal it doesn't really matter)

**Comment by chancancode** [comment]
This is awesome. Very happy with the way it is factored! 💯 

(Also, thank you for taking the time to write those tests, really appreciate it!)

File: packages/glimmer-runtime/lib/vm/update.ts (hunk 8 above)
Comment (krisselden) [inline packages/glimmer-runtime/lib/vm/update.ts]: tslint in now part of CI so that should fail the build.
